commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java

```diff
@@ -0,0 +1,469 @@
+            if (candidates.size() < 4) {
+                return this;
+            }
+
+            final List<Vector2D> quadrilateral = buildQuadrilateral(minY, maxX, maxY, minX);
+            // if the quadrilateral is not well formed, e.g. only 2 points, do not attempt to reduce
+            if (quadrilateral.size() < 3) {
+                return this;
+            }
+
+            // check all points if they are within the quadrilateral
+            // in which case they can not be part of the convex hull
+            if (!insideQuadrilateral(point, quadrilateral)) {
+                candidates.add(point);
+            }
+
+            return this;
+        }
+
+        /** Appends the given points to a collection of possible hull points, if and only
+         * if the given points are outside of a constructed quadrilateral of extreme
+         * properties.
+         *
+         * @param points a given collection of points.
+         * @throws NullPointerException if points is {@code null}.
+         * @return this instance.
+         */
+        public Builder append(Collection<Vector2D> points) {
+            points.forEach(this::append);
+            return this;
+        }
+
+        /**
+         * Build a convex hull from the set appended points.
+         *
+         * @return the convex hull
+         * @throws IllegalStateException if generator fails to generate a convex hull for
+         *      the given set of input points
+         */
+        public ConvexHull2D build() {
+            Collection<Vector2D> hullVertices;
+            if (candidates.size() < 2) {
+                hullVertices = candidates;
+            } else {
+                hullVertices = findHullVertices(candidates);
+            }
+
+            if (!isConvex(hullVertices)) {
+                throw new IllegalStateException("Convex hull algorithm failed to generate solution");
+            }
+
+            return new ConvexHull2D(hullVertices, precision);
+        }
+
+        /** Build the convex quadrilateral with the found corner points (with min/max x/y
+         * coordinates).
+         *
+         * @param points the respective points with min/max x/y coordinate
+         * @return the quadrilateral
+         */
+        private static List<Vector2D> buildQuadrilateral(final Vector2D... points) {
+            final List<Vector2D> quadrilateral = new ArrayList<>();
+            for (final Vector2D p : points) {
+                if (!quadrilateral.contains(p)) {
+                    quadrilateral.add(p);
+                }
+            }
+            return quadrilateral;
+        }
+
+        /** Checks if the given point supersedes one of the corners. If it does the old
+         * corner is removed and the point added to the collection of points.
+         *
+         * @param point a given point.
+         */
+        private void checkCorners(Vector2D point) {
+            if (minX == null || point.getX() < minX.getX()) {
```

Comment:
Here we only require one null check. If minX is null then the other corners are also null and all can be set to the point:
```Java
if (minX == null) {
    minX = maxX = minY = maxY = point;
    candidates.add(point);
    return;
}
```
IIUC this condition will also be met if `candidate.isEmpty()`. However using the null check does indicate that for the rest of the method the corners are not null.

Reply:
The observation is correct I changed the code. Although the method becomes a bit longer this way we can save 3 of the four null checks as suggested.